File: typescript/kata-03-wrap-method/src/employeeService.ts
// Legacy HR system code.
//
// This module is intentionally hard to test:
//  - EmployeeRepository is a SINGLETON whose getInstance() opens a real
//    database connection and whose methods issue real SQL.
//  - EmployeeService reaches for that singleton directly inside its methods,
//    so there is no seam to inject a fake repository.
//
// Do NOT refactor the repository into an injected dependency for this kata.
// The point is to add new behaviour using the Wrap Method technique without
// touching the existing, untested business logic.

export type EmployeeRecord = {
  id: string;
  name: string;
  title: string;
  level: number;
  yearsInRole: number;
  lastReviewScore: number; // 1..5
  onPerformanceImprovementPlan: boolean;
};

export type PromotionEvent = {
  type: 'employee.promoted';
  employeeId: string;
  toTitle: string;
  newLeadership: boolean;
};

export class PromotionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'PromotionError';
    Object.setPrototypeOf(this, PromotionError.prototype);
  }
}

// Publishes domain events. Two implementations exist: RealEventBus (production —
// its `publish` performs a genuine side effect) and, under `tests/`, an
// ObservableEventBus that merely records what was published so tests can assert
// on it — no database, no mocking framework.
//
// Inject a RealEventBus in the wrapper and an ObservableEventBus in the tests of
// your new publishing method.
export interface EventBus {
  publish(event: unknown): void;
}

// Production EventBus. In a real system `publish` would push to a message broker
// / outbox — a genuine, irreversible side effect.
//
// Here it deliberately throws: if this runs inside a test, you accidentally
// triggered a real publication. Inject an ObservableEventBus in tests instead,
// and keep this one for the wrapper's production wiring.
export class RealEventBus implements EventBus {
  publish(event: unknown): void {
    throw new Error(
      'RealEventBus.publish performed a REAL publication (message broker / subscribers). ' +
        'This must never run in a test — inject an ObservableEventBus and assert on ' +
        'its publishedEvents() instead.',
    );
  }
}

// Hard-to-test singleton: getInstance() and its methods clearly hit a real DB.
export class EmployeeRepository {
  private static instance: EmployeeRepository | null = null;

  private constructor() {
    // Opens a real database connection on first construction.
    // e.g. this.connection = mysql.createConnection(process.env.HR_DB_URL!);
    console.log('[EmployeeRepository] opening live database connection...');
  }

  static getInstance(): EmployeeRepository {
    if (EmployeeRepository.instance === null) {
      EmployeeRepository.instance = new EmployeeRepository();
    }
    return EmployeeRepository.instance;
  }

  findById(employeeId: string): EmployeeRecord {
    // Real query: SELECT * FROM employees WHERE id = ?
    throw new Error(
      `[EmployeeRepository] DB not available: cannot SELECT employee ${employeeId}`,
    );
  }

  save(record: EmployeeRecord): void {
    // Real query: UPDATE employees SET ... WHERE id = ?
    throw new Error(
      `[EmployeeRepository] DB not available: cannot UPDATE employee ${record.id}`,
    );
  }
}

export class EmployeeService {
  promote(employeeId: string, newTitle: string): void {
    const repo = EmployeeRepository.getInstance();
    const employee = repo.findById(employeeId);

    // Hardcoded promotion rules.
    if (employee.onPerformanceImprovementPlan) {
      throw new PromotionError(
        `Employee ${employeeId} is on a performance improvement plan and cannot be promoted`,
      );
    }
    if (employee.yearsInRole < 1) {
      throw new PromotionError(
        `Employee ${employeeId} must spend at least 1 year in role before promotion`,
      );
    }
    if (employee.lastReviewScore < 3) {
      throw new PromotionError(
        `Employee ${employeeId} needs a review score of at least 3 to be promoted`,
      );
    }
    if (newTitle.trim() === '' || newTitle === employee.title) {
      throw new PromotionError(
        `Invalid new title "${newTitle}" for employee ${employeeId}`,
      );
    }

    employee.title = newTitle;
    employee.level = employee.level + 1;
    employee.yearsInRole = 0;

    repo.save(employee);
  }

  demote(employeeId: string, newTitle: string): void {
    const repo = EmployeeRepository.getInstance();
    const employee = repo.findById(employeeId);

    // Hardcoded demotion rules.
    if (employee.level <= 1) {
      throw new PromotionError(
        `Employee ${employeeId} is already at the lowest level and cannot be demoted`,
      );
    }
    if (newTitle.trim() === '' || newTitle === employee.title) {
      throw new PromotionError(
        `Invalid new title "${newTitle}" for employee ${employeeId}`,
      );
    }

    employee.title = newTitle;
    employee.level = employee.level - 1;
    employee.yearsInRole = 0;

    repo.save(employee);
  }
}

File: java/kata-03-wrap-method/src/main/java/tech/qmates/kata/hr/EmployeeService.java
package tech.qmates.kata.hr;

/**
 * Legacy HR service. Both {@link #promote(String, String)} and
 * {@link #changeDepartment(String, String)} reach out to the
 * {@link EmployeeRepository} singleton directly, load a record, enforce a set
 * of hardcoded business rules, then persist the mutated record back through the
 * same singleton.
 */
public class EmployeeService {

    public void promote(String employeeId, String newTitle) throws PromotionException {
        EmployeeRepository repository = EmployeeRepository.getInstance();

        EmployeeRecord employee = repository.findById(employeeId);
        if (employee == null) {
            throw new PromotionException("Unknown employee: " + employeeId);
        }

        // Hardcoded business rules.
        if (newTitle == null || newTitle.isBlank()) {
            throw new PromotionException("A new title is required to promote an employee");
        }
        if (newTitle.equalsIgnoreCase(employee.getTitle())) {
            throw new PromotionException("Employee already holds the title: " + newTitle);
        }
        if (employee.getYearsInRole() < 1) {
            throw new PromotionException("Employee must spend at least 1 year in role before promotion");
        }
        if (employee.getPerformanceRating() < 3) {
            throw new PromotionException("Performance rating too low for promotion");
        }
        if (newTitle.startsWith("VP") && employee.getSalary() < 90_000) {
            throw new PromotionException("VP promotions require a base salary of at least 90000");
        }

        // Apply the promotion.
        employee.setTitle(newTitle);
        employee.setYearsInRole(0);
        employee.setSalary((int) Math.round(employee.getSalary() * 1.10));

        repository.save(employee);
    }

    public void changeDepartment(String employeeId, String newDepartment) throws PromotionException {
        EmployeeRepository repository = EmployeeRepository.getInstance();

        EmployeeRecord employee = repository.findById(employeeId);
        if (employee == null) {
            throw new PromotionException("Unknown employee: " + employeeId);
        }

        // Hardcoded business rules.
        if (newDepartment == null || newDepartment.isBlank()) {
            throw new PromotionException("A target department is required");
        }
        if (newDepartment.equalsIgnoreCase(employee.getDepartment())) {
            throw new PromotionException("Employee already works in: " + newDepartment);
        }
        if (employee.getYearsInRole() < 1) {
            throw new PromotionException("Employee must spend at least 1 year in role before transferring");
        }

        // Apply the transfer.
        employee.setDepartment(newDepartment);
        employee.setYearsInRole(0);

        repository.save(employee);
    }
}
